cmd/artifact/main.go

```diff
@@ -0,0 +1,341 @@
+			return cli.NewExitError("Must specify command", ErrBadUsage)
+		}
+		return cli.NewExitError(fmt.Sprintf("%s is not a command", c.Args().Get(0)), ErrBadUsage)
+	}
+
+	app.OnUsageError = func(context *cli.Context, err error, isSubcommand bool) error {
+		return cli.NewExitError(err, ErrBadUsage)
+	}
+
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:   "client-id",
+			EnvVar: "TASKCLUSTER_CLIENT_ID",
+			Usage:  "set client id to `CLIENT_ID`",
+		},
+		cli.StringFlag{
+			Name:   "access-token",
+			EnvVar: "TASKCLUSTER_ACCESS_TOKEN",
+			Usage:  "set access token to `ACCESS_TOKEN`",
+		},
+		cli.StringFlag{
+			Name:   "certificate",
+			EnvVar: "TASKCLUSTER_CERTIFICATE",
+			Usage:  "set certificate to `CERTIFICATE`",
+		},
+		cli.StringFlag{
+			Name: "base-url",
+			//EnvVar: "QUEUE_BASE_URL",
+			Usage: "set queue's `BASE_URL`",
+			Value: queue.DefaultBaseURL,
+		},
+		cli.IntFlag{
+			Name:   "chunk-size",
+			Usage:  "set the I/O chunk size to `CHUNK_SIZE` KB",
+			Value:  artifact.DefaultChunkSize / 1024,
+			EnvVar: "ARTIFACT_CHUNK_SIZE",
+		},
+		cli.IntFlag{
+			Name:   "part-size",
+			Usage:  "set the I/O chunk size to `PART_SIZE` MB",
+			Value:  artifact.DefaultPartSize * artifact.DefaultChunkSize / 1024 / 1024,
+			EnvVar: "ARTIFACT_PART_SIZE",
+		},
+		cli.BoolFlag{
+			Name:  "quiet, q",
+			Usage: "supress debugging output",
+		},
+		cli.BoolFlag{
+			Name:  "allow-insecure-requests",
+			Usage: "allow insecure (http) requests. NOT RECOMMENDED",
+		},
+	}
+
+	app.Commands = []cli.Command{
+		{
+			Name:    "download",
+			Aliases: []string{"d"},
+			Usage:   "download an artifact",
+			Flags: []cli.Flag{
+				cli.StringFlag{
+					Name:   "output, o",
+					Usage:  "`FILENAME` to write output to.  If not provided, Standard Output will be assumed",
+					EnvVar: "ARTIFACT_OUTPUT",
+				},
+				cli.BoolFlag{
+					Name:  "latest",
+					Usage: "request artifact from latest run",
+				},
+			},
+			ArgsUsage: "taskId runId name",
+			Action: func(c *cli.Context) error {
+				var err error
+				q, err := queue.New(&tcclient.Credentials{
+					ClientID:    c.GlobalString("client-id"),
+					AccessToken: c.GlobalString("access-token"),
+					Certificate: c.GlobalString("certificate"),
+				})
+				if err != nil {
+					return cli.NewExitError(err, ErrBadUsage)
+				}
+
+				if c.GlobalIsSet("base-url") {
+					q.BaseURL = c.GlobalString("base-url")
+				}
+
+				client := artifact.New(q)
+
+				if c.GlobalBool("allow-insecure-requests") {
+					client.AllowInsecure = true
+				}
+
+				if c.GlobalBool("quiet") {
+					artifact.SetLogOutput(ioutil.Discard)
+				}
+
+				if c.GlobalIsSet("chunk-size") {
+					cz := c.GlobalInt("chunk-size")
+					_, ps := client.GetInternalSizes()
+					err := client.SetInternalSizes(cz*1024, ps)
+					if err != nil {
+						return cli.NewExitError(err, ErrBadUsage)
+					}
+				}
+
+				if !c.IsSet("output") {
+					return cli.NewExitError("must specify output", ErrBadUsage)
```

Comment:
I've changed it so that you have to use "-" explicitly like many other unix utilities.  I suspect that's nicer than getting whole artifacts accidentally printed to logs by forgetting the output parameter

Reply:
actually, stdout doesn't support seeking even though it is a `os.File*`, which causes things to go badly.  Since I'm using type assertions to see if the output `io.Writer` is in fact also an `io.Seeker`, but Stdout can't seek, it's sort of broken.

I doubt writing to stdout is worthwihle?

Reply:
hmm, i could probably check if output == os.Stdout... I dunno.

Reply:
In the end, I decided that because these are extra checks, I'll instead try to seek, and only if the seek succeeds and the filesize isn't == 0 throw an error.